protocols/relay/src/transport.rs

```diff
@@ -0,0 +1,432 @@
+    report_listen_addr: Option<Multiaddr>,
+}
+
+impl<T: Transport> Stream for RelayListener<T> {
+    type Item = Result<
+        ListenerEvent<RelayedListenerUpgrade<T>, EitherError<<T as Transport>::Error, RelayError>>,
+        EitherError<<T as Transport>::Error, RelayError>,
+    >;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+
+        if let Some(msg) = this.msg_to_behaviour {
+            match Future::poll(msg.as_mut(), cx) {
+                Poll::Ready(Ok(())) => *this.msg_to_behaviour = None,
+                Poll::Ready(Err(e)) => return Poll::Ready(Some(Err(EitherError::B(e.into())))),
+                Poll::Pending => {}
+            }
+        }
+
+        if let Some(listener) = this.inner_listener.as_pin_mut() {
+            match listener.poll_next(cx) {
+                Poll::Ready(Some(Err(e))) => return Poll::Ready(Some(Err(EitherError::A(e)))),
+                Poll::Ready(Some(Ok(ListenerEvent::Upgrade {
+                    upgrade,
+                    local_addr,
+                    remote_addr,
+                }))) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::Upgrade {
+                        upgrade: RelayedListenerUpgrade::Inner(upgrade),
+                        local_addr,
+                        remote_addr,
+                    })))
+                }
+                Poll::Ready(Some(Ok(ListenerEvent::NewAddress(addr)))) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(addr))))
+                }
+                Poll::Ready(Some(Ok(ListenerEvent::AddressExpired(addr)))) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::AddressExpired(addr))))
+                }
+                Poll::Ready(Some(Ok(ListenerEvent::Error(err)))) => {
+                    return Poll::Ready(Some(Ok(ListenerEvent::Error(EitherError::A(err)))))
+                }
+                Poll::Ready(None) => return Poll::Ready(None),
+                Poll::Pending => {}
+            }
+        }
+
+        match this.from_behaviour.lock().unwrap().poll_next_unpin(cx) {
+            Poll::Ready(Some(BehaviourToTransportMsg::IncomingRelayedConnection {
+                stream,
+                src_peer_id,
+                relay_addr,
+            })) => {
+                return Poll::Ready(Some(Ok(ListenerEvent::Upgrade {
+                    upgrade: RelayedListenerUpgrade::Relayed(Some(stream)),
+                    local_addr: relay_addr.with(Protocol::P2pCircuit),
+                    remote_addr: Protocol::P2p(src_peer_id.into()).into(),
+                })));
+            }
+            Poll::Ready(None) => unimplemented!(),
+            Poll::Pending => {}
+        }
+
+        // TODO: This should likely only be reported once the relay connection is established. But
+        // how should one know that?
```

Comment:
Sounds like one would need a dedicated channel between the behaviour and this particular listener? I'm wondering if the `to_transport` sender of such a fresh channel could not be sent to the behaviour in the `TransportToBehaviourMsg::ListenReq` for a `listen_on()`, thereby removing the `Arc<Mutex>` in the `RelayTransportWrapper` and maybe solving the problem of emitting `IncomingDstReqSuccess` to the right listener? Apologies if this is a naive suggestion, I'm still wrapping my head around the details.

Reply:
> Apologies if this is a naive suggestion, I'm still wrapping my head around the details.

No, I agree that this is the way to go, only issue I have today is how to support use-case (2) (an incoming relayed connection via a relay node one is not explicitly listening via) (see https://github.com/libp2p/rust-libp2p/pull/1838#discussion_r586447824).